src_feedback/lib.rs

```diff
@@ -0,0 +1,892 @@
+use ksym::{KSymResolver, KSymCache};
+use elf_cache::{ElfBackend, ElfCache};
+
+struct CacheHolder {
+    ksym: KSymCache,
+    elf: ElfCache,
+}
+
+impl CacheHolder {
+    fn new() -> CacheHolder {
+	CacheHolder {
+	    ksym: ksym::KSymCache::new(),
+	    elf: elf_cache::ElfCache::new(),
+	}
+    }
+
+    fn get_ksym_cache(&self) -> &KSymCache {
+	&self.ksym
+    }
+
+    fn get_elf_cache(&self) -> &ElfCache {
+	&self.elf
+    }
+}
+
+pub trait StackFrame {
+    fn get_ip(&self) -> u64;
+    fn get_frame_pointer(&self) -> u64;
+}
+
+pub trait StackSession {
+    fn next_frame(&mut self) -> Option<&dyn StackFrame>;
+    fn prev_frame(&mut self) -> Option<&dyn StackFrame>;
+    fn go_top(&mut self);
+}
+
+pub struct AddressLineInfo {
+    pub path: String,
+    pub line_no: usize,
+    pub column: usize,
+}
+
+/// The trait of symbol resolvers.
+///
+/// An symbol resolver usually provides information from one symbol
+/// source; e., a symbol file.
+pub trait SymResolver {
+    /// Return the range that this resolver serve in an address space.
+    fn get_address_range(&self) -> (u64, u64);
+    /// Find the name and the start address of a symbol found for
+    /// the given address.
+    fn find_symbol(&self, addr: u64) -> Option<(&str, u64)>;
+    /// Find the address of a symbol anme.
+    fn find_address(&self, name: &str) -> Option<u64>;
+    /// Find the file name and the line number of an address.
+    fn find_line_info(&self, addr: u64) -> Option<AddressLineInfo>;
+
+    fn repr(&self) -> String;
+}
+
+pub const REG_RAX: usize = 0;
+pub const REG_RBX: usize = 1;
+pub const REG_RCX: usize = 2;
+pub const REG_RDX: usize = 3;
+pub const REG_RSI: usize = 4;
+pub const REG_RDI: usize = 5;
+pub const REG_RSP: usize = 6;
+pub const REG_RBP: usize = 7;
+pub const REG_R8: usize = 8;
+pub const REG_R9: usize = 9;
+pub const REG_R10: usize = 10;
+pub const REG_R11: usize = 11;
+pub const REG_R12: usize = 12;
+pub const REG_R13: usize = 13;
+pub const REG_R14: usize = 14;
+pub const REG_R15: usize = 15;
+pub const REG_RIP: usize = 16;
+
+struct X86_64StackFrame {
+    rip: u64,
+    rbp: u64,
+}
+
+impl StackFrame for X86_64StackFrame {
+    fn get_ip(&self) -> u64 {
+	self.rip
+    }
+    fn get_frame_pointer(&self) -> u64 {
+	self.rbp
+    }
+}
+
+/// Do stacking unwind for x86_64
+///
+/// Parse a block of memory that is a copy of stack of thread to get frames.
+///
+pub struct X86_64StackSession {
+    frames: Vec<X86_64StackFrame>,
+    stack: Vec<u8>,
+    stack_base: u64,		// The base address of the stack
+    registers: [u64; 17],
+    current_rbp: u64,
+    current_rip: u64,
+    current_frame_idx: usize,
+}
+
+impl X86_64StackSession {
+    fn _get_rbp_rel(&self) -> usize {
+	(self.current_rbp - self.stack_base) as usize
+    }
+
+    fn _mark_at_bottom(&mut self) {
+	self.current_rbp = 0;
+    }
+
+    fn _is_at_bottom(&self) -> bool {
+	self.current_rbp == 0
+    }
+
+    fn _get_u64(&self, off: usize) -> u64 {
+	let stack = &self.stack;
```

Comment:
Do we have to account for endianness here? 

btw, the u64 primitive in Rust has the from_ne/le/be_bytes which takes in a byte array and returns a u64 (https://doc.rust-lang.org/std/primitive.u64.html#method.from_ne_bytes)

Reply:
I did think about endianness at the beginning. That is why I don't do casting directly.  As you mentioned, there must be some better way to do that.  I focused on prototyping, so I added a separate function like this without killing too many brain cells.

Anyway, it is an excellent time to work on the endianness now since we will work on some applications of BlazeSym.
You can send PR to me if you like.  I believe there is a bunch of things to improve.